### PR TITLE
Bootstrapify gallery plugin

### PR DESCRIPTION
**Pull request by Bakual**
Since it was suggested in various comments so far to get rid of the fotorama JS and CSS in the custom gallery field

### Summary of Changes
Removes Fotorama media and changes the gallery plugin to use Bootstrap 2.3 thumbnails.
It adds also a new option to specify how many thumbnails it should show per row. This is limited to values that make sense in a Bootstrap scenario (1,2,3,4,6,12).

### Testing Instructions
* Set up a custom field of type gallery and assign to an article
* Check frontend

### Expected result
bootstrap

### Actual result
fotorama

### Documentation Changes Required
If there is already a documentation about this custom field, then it needs to be adjusted.

### CODE REVIEW

**Comment by Bakual** [comment]
@dgt41 @coolcat-creations This one's for you two 😄

**Comment by laoneo** [comment]
Can he images made somehow with equal height?
image

Also when I click on an image it opens the page with only the image. Perhaps just open them in a modal window (for the record, I'm not a fan of modals).

File: plugins/fields/gallery/params/gallery.xml [hunk 1]
```diff
@@ -27,6 +28,21 @@
 			/>
 
 			<field
+				name="thumbs_row"
+				type="list"
+				label="PLG_FIELDS_GALLERY_PARAMS_THUMBS_ROW_LABEL"
+				description="PLG_FIELDS_GALLERY_PARAMS_THUMBS_ROW_DESC"
+				default="6"
+				>
+				<option value="1">1</option>
```
Comment (zero-24) [inline plugins/fields/gallery/params/gallery.xml]: should we not use: https://github.com/joomla/joomla-cms/blob/staging/administrator/language/en-GB/en-GB.ini#L12-L33 ?

**Comment by Bakual** [comment]
The thumbnails are done using the default Bootstrap thumbnails component (http://getbootstrap.com/2.3.2/components.html#thumbnails).
I didn't touch the picture resize stuff you had put into the original plugin, so maybe that can be used to make them equal height? I haven't tested that as I assumed it works 😄 

> Also when I click on an image it opens the page with only the image. Perhaps just open them in a modal window (for the record, I'm not a fan of modals).

That should be doable, yes. But don't ask me how to do that without already loading all the full-resolution images. Maybe someone else with more knowledge in Bootstrap modals knows that.

File: plugins/fields/gallery/params/gallery.xml (hunk 1 above)
Comment (Bakual) [inline plugins/fields/gallery/params/gallery.xml]: Probably, it misses the 12 though.
Will look at that tomorrow.

Comment (zero-24) [inline plugins/fields/gallery/params/gallery.xml]: Than lets add the missing values between 10 - 15 ;)

**Comment by laoneo** [comment]
The thumbnail code worked for fotorama, you can't expect that it works for your code as well.

There is no need to load the full sized image as well, you can just pass the full url to the modal window. If you can't figure it out, I can have a look on Monday.

I would not use the bs modal, too many conflicts with bs3 templates, but the joomla modal should work here.

**Comment by Bakual** [comment]
>The thumbnail code worked for fotorama, you can't expect that it works for your code as well.

Honestly I haven't looked at where exactly it is used. I just saw there is some resizing done in PHP and assumed it has to do with that. That part should work independant of the Fotorama JS as it is done in PHP.

> There is no need to load the full sized image as well, you can just pass the full url to the modal window. If you can't figure it out, I can have a look on Monday.

Yep, that's the part I have not figured out yet.

> I would not use the bs modal, too many conflicts with bs3 templates, but the joomla modal should work here.

Doesn't the Joomla modal use Bootstrap?

File: plugins/fields/gallery/params/gallery.xml (hunk 1 above)
Comment (Bakual) [inline plugins/fields/gallery/params/gallery.xml]: Done 👍

**Comment by Bakual** [comment]
I removed the maxWidth and maxHeight settings and related codde as they indeed are useless now as you said.
The only resizing that is done here is for the thumbnails.
As for equal height, currently the code makes it equal width. We could make it optional do it equal height, but imho width will work better (Bootstrap woudl shrink the image if it's to wide otherwise).
Your example screenshot probably was  a bit  a theoretical case as usually in galleries you will have most likely pictures of similar formats, not banner and photos mixed. You may get landscape and portrait formats mixed thought.

**Comment by zero-24** [comment]
@Bakual please add the new strings also to the frontend en-GB ini file ;)

**Comment by Bakual** [comment]
😊

**Comment by infograf768** [comment]
Works here, but we indeed need a modal to not be forced to use browser back in history.

**Comment by infograf768** [comment]
Also, maybe add .jpeg here
`$filter = '\.jpg$|\.png$|\.bmp$|\.gif$';`

**Comment by Bakual** [comment]
Added jpeg to file filter.
As for the modal I agree, but need some guidance. If we have a modal with URL injection already somewhere in core just point me to it and I'll figure it out.

**Comment by Bakual** [comment]
Changed so image opens in a modal. However I think that has to be done in a better way.
* I had to create a modal for each image which is ugly when looking at the HTML code. The full image is only loaded when the modal is openend, but still. Imho there must be a way where the URL is part of the modal call itself (using data attributes) and the modal needs to be specified only once.
* While I was able to dynamically specify the height of the modal based on the image properties, the modalWidth can't be set in pixels, only in vh to my knowledge. Not sure what the best approach is. Currently the image will be left aligned in the modal with a big white empty space to the right, which looks bad.
* Having a title and footer for the modal is a bit pointless for an image. But at least the title part is always there even if no title is specified. I haven't tried yet to omit the footer as it was my least concern :smile: 

I'm thinking the modal approach isn't the best choice here. Anyone has a better idea or knows how the modal has to be used properly here?

**Comment by Bakual** [comment]
Please disregard the previous comment. Due to the issues I wrote I changed it to a popover approach.
Now the original image will open in a popover when hovered.

I will leave it like that until someone can provide a better way.

**Comment by infograf768** [comment]
will test asap as changes in lang strings

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 2d218a8a012306bf8d1109b512a4566b5b31017a<br /><br />@Bakual the original Size isn't openend:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14238">issues.joomla.org/tracker/joomla-cms/14238</a>.</sub>

**Comment by Bakual** [comment]
I see, apparently the popover does some resizing using CSS. Which imho actually is a good thing as big pictures will be resized.
For me it is fine as it is. If someone has a better idea I'm open for code suggestions :smile:

**Comment by infograf768** [comment]
Honestly, this is no good. It makes the gallery quite useless.
What we need is a lightbox. It exists for bootsrap, but needs BS >=3.10

**Comment by infograf768** [comment]
I added in Protostar CSS:
```css
.popover{
    max-width: 600px; 
}
```
and I can get

screen shot 2017-03-01 at 17 55 32

screen shot 2017-03-01 at 17 57 31

**Comment by zero-24** [comment]
@bakual drone is ailing on a missing space ;)

**Comment by Bakual** [comment]
@zero-24 fixed 👍

**Comment by Bakual** [comment]
> Honestly, this is no good. It makes the gallery quite useless. 
What we need is a lightbox. It exists for bootsrap, but needs BS >=3.10

Yep, there is no lightbox in core, and I don't want to add one.
Keep also in mind that the existing gallery also has no lightbox. It just shows a bigger image above the thumbnails. So we don't really **need** one.

As for the CSS, I would leave that to the site integrators / template designers. I don't want to add CSS in this custom form field.
It depends a lot on the usecase of that gallery plugin. For some cases this proposal is fine, for others they need a bigger one or even a lightbox. The nice thing is that it is completely overrideable and customisable.

**Comment by infograf768** [comment]
It does not cost much to add the max-width param.
if no figure is entered in the field or if figure == 270px, no css would be added and anyone can do whatever they like.
For Protostar and similar templates (many users do NOT use complex "integrators" templates), the max-width makes a lot of sense.
We just have to explain in the tip for that field what it would do.

The existing state of the gallery as well as this proposal, without at least the choice **for the users who need this max-width** is totally useless.
Basically, as the code would be in the tmpl, it will remain overrideable and customisable.

 In that case, let's get rid of this plugin and let people create specific plugins who will really be useful.

**Comment by dgt41** [comment]
> In that case, let's get rid of this plugin and let people create specific plugins who will really be useful.

Excellent idea! :)

**Comment by infograf768** [comment]
I tested this:
(BTW, in this PR,  there are obsolete fields in /plugins/fields/gallery/gallery.xml i.e. not equal to //plugins/fields/gallery/params/gallery.xml)

Adding a field
```xml
<field
				name="popover_maxwidth"
				type="integer"
				label="PLG_FIELDS_GALLERY_PARAMS_POPOVER_MAXWIDTH_LABEL"
				description="PLG_FIELDS_GALLERY_PARAMS_POPOVER_MAXWIDTH_WIDTH_DESC"
				first="270"
				last="2000"
				step="10"
			/>
```

then modified the tmpl /plugins/fields/gallery/tmpl/gallery.php by adding
```php
$popoverMaxWidth = $fieldParams->get('popover_maxwidth');
$thumbWidth      = $fieldParams->get('thumbnail_width', '64');
$thumbsRow       = $fieldParams->get('thumbs_row', 6);
$spanClass       = 'span' . 12 / $thumbsRow;

if ($popoverMaxWidth && $popoverMaxWidth !== 270)
{
	$doc->addStyleDeclaration('
		.popover {
			max-width: ' . $popoverMaxWidth . 'px;
		}
	');
}
```
whch does the job.

**Comment by Bakual** [comment]
Your proposal would change the max-width of each popovr used in core when that field is active.
Also you can't add CSS to the head like that with a custom field that may be present multiple times. You'd end up with multiple such declarations..
Both are things I don't want to do in a custom field. Sorry.

**Comment by infograf768** [comment]
OK, then let's kill this field and be over with.

**Comment by laoneo** [comment]
Joomla comes with Squeezebox, something like that should work:

	var width = jQuery(window).width();
	var height = jQuery(window).height();
	SqueezeBox.open(url, {
		handler: 'iframe',
		size: { x: width < 650 ? width - (width * 0.10) : 650, y: height < 650 ? height - (height * 0.10) : 650 }
	});

Just when you guys still wanne use a modal. But like it is now, I would also skip it honestly.

**Comment by infograf768** [comment]
Made a PR to remove this plugin from core
https://github.com/joomla/joomla-cms/pull/14309

**Comment by dgt41** [comment]
@laoneo that's a huge no from me as `SqueezeBox` will bring back Mootools on the front end

**Comment by laoneo** [comment]
@dgt41 you will see in a few years, mootools will be again state of the art and "The" javascript library you want to use in your project ;-)

**Comment by dgt41** [comment]
That's a reason that I don't want to use any of the frameworks out there. Their lifespan is given (jQuery included)

**Comment by zero-24** [comment]
Field is removed now from the core

**Comment by coolcat-creations** [comment]
Sorry i somehow missed this topic... Suggestion: Can we create somewhere a small library of plugins or extensions that never made it in the core but are still useful or a good base? My heart is bleeding when work that is done is thrown away.

**Comment by zero-24** [comment]
What about joomla-extensions ?

**Comment by Bakual** [comment]
I'm quite sure the JED will be filled up with such plugins fast. They probably need to create a new category thought.
I can save my work and put it to my repo if that helps you.

**Comment by infograf768** [comment]
if we get boostrap 3.x in j 4, this plugin can easily be used again with little change

**Comment by dgt41** [comment]
@infograf768 it should be bootstrap 4, but the changes are still an easy task

**Comment by Bakual** [comment]
@coolcat-creations See https://github.com/Bakual/CustomFieldGallery
Download it as zipfile and you can directly install it. Should be same code.

**Comment by coolcat-creations** [comment]
@Bakual thank you!!
